File: shared/src/domain/scheme.ts
import type { ISODateString, LanguageCode, SchemeCode, StateCode, UUID } from "./common";

export type SchemeCategory =
  | "education"
  | "health"
  | "housing"
  | "employment"
  | "agriculture"
  | "pension"
  | "women"
  | "disability"
  | "food-security"
  | "financial-inclusion"
  | "other";

/** Central = pan-India; state = a specific state's scheme. */
export type SchemeScope = "central" | "state";

/**
 * Publication lifecycle of a scheme (mirrors the `schemes.scheme_status` column).
 * Only `published` appears in the public catalog; the other statuses drive the
 * admin content+pipeline (draft → pending_review → verified → published).
 */
export type SchemeStatus =
  | "draft"
  | "pending_review"
  | "verified"
  | "published"
  | "temporarily_unavailable"
  | "archived"
  | "expired";

/** Content source/verification state shown to admins (Prompt 15). */
export type SchemeVerificationStatus =
  | "unverified"
  | "pending"
  | "verified"
  | "failed"
  | "stale";

/** Accepted sort keys for the schemes list endpoint. */
export type SchemeSort = "relevance" | "updated" | "popular";

/** A step in the application/renewal process. */
export interface SchemeApplicationStep {
  step: number;
  title: LocalizedText;
  description?: LocalizedText;
  /** Online / offline / both — which channel the step is done through. */
  mode?: "online" | "offline" | "both";
}

/** Frequently-asked question attached to a scheme. */
export interface SchemeFAQ {
  id: UUID;
  question: LocalizedText;
  answer: LocalizedText;
}

/** A saved (bookmarked) search the user wants to revisit. */
export interface SavedSearch {
  id: UUID;
  query: string;
  filters?: Record<string, string | boolean | number | undefined>;
  notifyOnUpdate: boolean;
  createdAt: ISODateString;
}

/** One entry in the user's recent-search history. */
export interface SearchHistoryItem {
  id: UUID;
  query: string;
  createdAt: ISODateString;
}

/** Autocomplete / popular-search suggestions for a partial query. */
export interface SearchSuggestions {
  query: string;
  /** Completion terms (words/tags/categories) for the prefix. */
  suggestions: string[];
  /** Closest known scheme/category name (for misspellings), if any. */
  corrected?: string;
}

/** Lightweight scheme in the trending/popular rails. */
export interface SchemeTrending {
  rank: number;
  scheme: SchemeSummary;
}

export interface SchemeSearchQuery {
  q?: string;
  category?: SchemeCategory;
  subCategory?: string;
  state?: StateCode;
  scope?: SchemeScope;
  ministry?: string;
  department?: string;
  /** Demographic filters evaluated against the scheme's eligibility rules. */
  gender?: string;
  occupation?: string;
  incomeBand?: string;
  education?: string;
  isFarmer?: boolean;
  isStudent?: boolean;
  isDisabled?: boolean;
  isMinority?: boolean;
  isSeniorCitizen?: boolean;
  isSelfEmployed?: boolean;
  isWomen?: boolean;
  page?: number;
  pageSize?: number;
  sort?: SchemeSort;
}

/** Field names the eligibility engine can evaluate against a UserProfile. */
export type EligibilityField =
  | "age"
  | "gender"
  | "state"
  | "district"
  | "income_band"
  | "caste_category"
  | "occupation"
  | "education"
  | "is_farmer"
  | "is_student"
  | "is_disabled"
  | "is_minority"
  | "is_senior_citizen"
  | "is_self_employed"
  | "is_widow"
  | "is_women";

export type EligibilityOperator = "eq" | "neq" | "gte" | "lte" | "in" | "between" | "exists";

export interface EligibilityRule {
  field: EligibilityField;
  operator: EligibilityOperator;
  value: string | number | [number, number] | string[];
  /** Human-readable explanation in English (translated on the client). */
  description: string;
}

/** Localized document with name/summary pair used across every scheme resource. */
export interface LocalizedText {
  en: string;
  native: string;
}

/** Canonical document category used for OCR + checklist grouping. */
export type DocumentKind =
  | "identity"
  | "address"
  | "income"
  | "age"
  | "caste"
  | "bank"
  | "land"
  | "disability"
  | "family"
  | "photo"
  | "other";

export interface RequiredDocument {
  id: UUID;
  /** Canonical document type, e.g. "AADHAAR", "RATION_CARD". */
  name: string;
  kind: DocumentKind;
  localizedNames: Partial<Record<LanguageCode, string>>;
  optional: boolean;
  /** Whether OCR (tesseract/PaddleOCR) can extract data from this doc. */
  ocrSupported: boolean;
  /** Localized guidance for obtaining/using this document. */
  verificationHint?: string;
  notes?: string;
}

export interface SchemeApplicationLinks {
  /** Official online application URL (post.onlinereg etc). */
  online?: string;
  /** Offline application description (CSC center, SDM office). */
  offline?: string;
  helpline?: string;
  /** Official portal source URL used for verification. */
  sourceUrl?: string;
}

/**
 * A government scheme (Prompt 6 schema). `scope: "central"` schemes apply to all
 * states (`applicableStates` empty means pan-India); `scope: "state"` schemes
 * must list their states. All rich text is bilingual (English + native) at
 * rest; other languages are translated on demand.
 */
export interface Scheme {
  id: UUID;
  code: SchemeCode;
  /** Compact display name, e.g. "PM-KISAN". */
  shortName?: string;
  name: LocalizedText;
  summary: LocalizedText;
  /** Full, structured description (covers Overview section). */
  description: LocalizedText;
  category: SchemeCategory;
  /** Optional sub-category, e.g. "scholarship" under "education". */
  subCategory?: string;
  ministry: string;
  /** Responsible department (for state schemes, the state department). */
  department?: string;
  scope: SchemeScope;
  /** Legacy single-state tag used by chat cards ("*" = central). */
  stateCode: StateCode;
  /** States the scheme applies to; empty array = all India. */
  applicableStates: StateCode[];
  /** Who the scheme is aimed at, e.g. ["landholding farmer families"]. */
  targetBeneficiaries: string[];
  benefits: string[];
  eligibilityRules: EligibilityRule[];
  requiredDocuments: RequiredDocument[];
  applicationSteps: SchemeApplicationStep[];
  /** How to renew (periodicity + process), if renewable. */
  renewalProcess?: LocalizedText;
  applicationLinks: SchemeApplicationLinks;
  officialWebsite?: string;
  officialApplicationLink?: string;
  helpline?: string;
  faqs: SchemeFAQ[];
  schemeStatus: SchemeStatus;
  /** Last time the content was verified against the official source. */
  lastVerifiedAt: string;
  /** Source provenance + verification state (Prompt 15 admin surface). */
  sourceName?: string;
  sourceUrl?: string;
  sourceType?: string;
  verificationStatus?: SchemeVerificationStatus;
  reviewNote?: string;
  /** Search keywords / synonyms beyond tags, e.g. ["kisan", "crop loan"]. */
  keywords: string[];
  /** Free-text tags for search, e.g. ["farmers", "KCC"]. */
  tags: string[];
  /** Global popularity index (used for trending/ordering). */
  popularity?: number;
  viewCount?: number;
  bookmarkCount?: number;
  validUntil?: string;
}

export interface SchemeSummary {
  id: UUID;
  code: SchemeCode;
  category: SchemeCategory;
  scope: SchemeScope;
  stateCode: StateCode;
  shortName?: string;
  name: LocalizedText;
  summary: LocalizedText;
  tags: string[];
  matchScore?: number;
  popularity?: number;
}

/** Paged list response envelope used by all list endpoints. */
export interface Paginated<T> {
  items: T[];
  page: number;
  pageSize: number;
  total: number;
}

File: server/app/schemas/scheme.py
"""Pydantic schemas for the scheme catalog + user scheme interactions.

Field names are camelCase over the wire (via ``APIModel``), mirroring
``shared/src/domain/scheme.ts``.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import Field, field_validator

from app.schemas.common import APIModel

SCHEME_CATEGORIES = (
    "education",
    "health",
    "housing",
    "employment",
    "agriculture",
    "pension",
    "women",
    "disability",
    "food-security",
    "financial-inclusion",
    "other",
)
SCHEME_SCOPES = ("central", "state")
SCHEME_STATUSES = (
    "draft",
    "pending_review",
    "verified",
    "published",
    "temporarily_unavailable",
    "archived",
    "expired",
)
#: Statuses considered "live" for the public catalog (search/browse/RAG).
PUBLIC_SCHEME_STATUSES = ("published",)
SCHEME_VERIFICATION_STATUSES = ("unverified", "pending", "verified", "failed", "stale")
SCHEME_SORTS = ("relevance", "updated", "popular")


class LocalizedText(APIModel):
    en: str = Field(default="", max_length=2000)
    native: str = Field(default="", max_length=2000)


class SchemeCreate(APIModel):
    """Admin create payload. Rich fields are validated as structured JSON."""

    code: str = Field(min_length=2, max_length=40, pattern=r"^[A-Z0-9][A-Z0-9._-]*$")
    short_name: str | None = Field(default=None, max_length=80)
    name: LocalizedText
    summary: LocalizedText
    description: LocalizedText
    category: str
    sub_category: str | None = Field(default=None, max_length=60)
    ministry: str = Field(default="", max_length=120)
    department: str | None = Field(default=None, max_length=120)
    scope: Literal["central", "state"] = "central"
    state_code: str = "*"
    applicable_states: list[str] = Field(default_factory=list, max_length=40)
    target_beneficiaries: list[str] = Field(default_factory=list, max_length=50)
    benefits: list[str] = Field(default_factory=list, max_length=100)
    eligibility_rules: list[dict[str, Any]] = Field(default_factory=list)
    required_documents: list[dict[str, Any]] = Field(default_factory=list)
    application_steps: list[dict[str, Any]] = Field(default_factory=list)
    renewal_process: dict[str, Any] | None = None
    application_links: dict[str, Any] = Field(default_factory=dict)
    official_website: str | None = Field(default=None, max_length=500)
    official_application_link: str | None = Field(default=None, max_length=500)
    faqs: list[dict[str, Any]] = Field(default_factory=list)
    keywords: list[str] = Field(default_factory=list, max_length=100)
    tags: list[str] = Field(default_factory=list, max_length=100)
    scheme_status: Literal[
        "draft",
        "pending_review",
        "verified",
        "published",
        "temporarily_unavailable",
        "archived",
        "expired",
    ] = "published"
    last_verified_at: datetime | None = None
    #: Source provenance + content verification (Prompt 15 admin).
    source_name: str | None = Field(default=None, max_length=200)
    source_url: str | None = Field(default=None, max_length=500)
    source_type: str | None = Field(default=None, max_length=40)
    verification_status: Literal["unverified", "pending", "verified", "failed", "stale"] = (
        "unverified"
    )
    review_note: str | None = Field(default=None, max_length=500)

    @field_validator("category")
    @classmethod
    def _category_known(cls, value: str) -> str:
        if value not in SCHEME_CATEGORIES:
            raise ValueError(f"Unknown category '{value}'.")
        return value

    @field_validator("applicable_states", "tags", "keywords", mode="before")
    @classmethod
    def _strip_blank(cls, value: Any) -> Any:
        if not isinstance(value, list):
            return value
        return [item.strip() for item in value if isinstance(item, str) and item.strip()]


class SchemeUpdate(APIModel):
    """Admin update payload — every field optional, partial merges only."""

    short_name: str | None = Field(default=None, max_length=80)
    name: LocalizedText | None = None
    summary: LocalizedText | None = None
    description: LocalizedText | None = None
    category: str | None = None
    sub_category: str | None = Field(default=None, max_length=60)
    ministry: str | None = Field(default=None, max_length=120)
    department: str | None = Field(default=None, max_length=120)
    scope: Literal["central", "state"] | None = None
    state_code: str | None = None
    applicable_states: list[str] | None = None
    target_beneficiaries: list[str] | None = None
    benefits: list[str] | None = None
    eligibility_rules: list[dict[str, Any]] | None = None
    required_documents: list[dict[str, Any]] | None = None
    application_steps: list[dict[str, Any]] | None = None
    renewal_process: dict[str, Any] | None = None
    application_links: dict[str, Any] | None = None
    official_website: str | None = Field(default=None, max_length=500)
    official_application_link: str | None = Field(default=None, max_length=500)
    faqs: list[dict[str, Any]] | None = None
    keywords: list[str] | None = None
    tags: list[str] | None = None
    scheme_status: (
        Literal[
            "draft",
            "pending_review",
            "verified",
            "published",
            "temporarily_unavailable",
            "archived",
            "expired",
        ]
        | None
    ) = None
    last_verified_at: datetime | None = None
    source_name: str | None = Field(default=None, max_length=200)
    source_url: str | None = Field(default=None, max_length=500)
    source_type: str | None = Field(default=None, max_length=40)
    verification_status: Literal["unverified", "pending", "verified", "failed", "stale"] | None = (
        None
    )
    review_note: str | None = Field(default=None, max_length=500)

    @field_validator("category")
    @classmethod
    def _category_known(cls, value: str | None) -> str | None:
        if value is not None and value not in SCHEME_CATEGORIES:
            raise ValueError(f"Unknown category '{value}'.")
        return value


class SchemeOut(APIModel):
    id: str
    code: str
    short_name: str | None = None
    name: LocalizedText
    summary: LocalizedText
    description: LocalizedText
    category: str
    sub_category: str | None = None
    ministry: str
    department: str | None = None
    scope: str
    state_code: str
    applicable_states: list[str] = Field(default_factory=list)
    target_beneficiaries: list[str] = Field(default_factory=list)
    benefits: list[str] = Field(default_factory=list)
    eligibility_rules: list[dict[str, Any]] = Field(default_factory=list)
    required_documents: list[dict[str, Any]] = Field(default_factory=list)
    application_steps: list[dict[str, Any]] = Field(default_factory=list)
    renewal_process: dict[str, Any] | None = None
    application_links: dict[str, Any] = Field(default_factory=dict)
    official_website: str | None = None
    official_application_link: str | None = None
    helpline: str | None = None
    faqs: list[dict[str, Any]] = Field(default_factory=list)
    scheme_status: str = "published"
    last_verified_at: datetime | None = None
    source_name: str | None = None
    source_url: str | None = None
    source_type: str | None = None
    verification_status: str = "unverified"
    review_note: str | None = None
    keywords: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    popularity: int = 0
    view_count: int = 0
    bookmark_count: int = 0
    created_at: datetime | None = None
    updated_at: datetime | None = None


class SchemeSummaryOut(APIModel):
    id: str
    code: str
    category: str
    scope: str
    state_code: str
    short_name: str | None = None
    name: LocalizedText
    summary: LocalizedText
    tags: list[str] = Field(default_factory=list)
    match_score: int | None = None
    popularity: int = 0


class SchemeSearchResultOut(SchemeSummaryOut):
    """Search result — summary plus a 0-100 relevance score."""


class SuggestionsOut(APIModel):
    query: str
    suggestions: list[str] = Field(default_factory=list)
    corrected: str | None = None


class TrendingOut(APIModel):
    rank: int
    scheme: SchemeSummaryOut


class BookmarkStatusOut(APIModel):
    scheme_id: str
    saved: bool
    bookmark_count: int = 0


class SavedSearchOut(APIModel):
    id: str
    query: str
    filters: dict[str, Any] = Field(default_factory=dict)
    notify_on_update: bool = False
    created_at: datetime


class SavedSearchCreate(APIModel):
    query: str = Field(min_length=1, max_length=200)
    filters: dict[str, Any] = Field(default_factory=dict)
    notify_on_update: bool = False


class SearchHistoryOut(APIModel):
    id: str
    query: str
    created_at: datetime
